### PR TITLE
FIR UAST: fixes about (implicit) constructors

### PR DESCRIPTION
**Pull request by jsjeon**
To be updated: first 12 commits are identical to #95

------

This PR includes small fixes regarding (implicit) constructors. Some of such fixes are in FIR IDE and LC sides.

This journey started from a difference (actually, lack of implicit constructor call) in a certain test. In `KotlinConstructorUMethod`'s body, we add an implicit constructor call if `KtSuperTypeCallEntry` is found in super type list. It turns out that `psi` origin is `null`ed out if the containing class has an implicit constructor only, hence no super type list. Allowing fake `ImplicitConstructor` source resolved the case.

Then, such call is added, but resolved call target---FIR constructor symbol---is not `Name`d. I revised the resolved function name API to render `<init>` for such case, but we could discuss alternatives.

Next, still some were missed when a secondary constructor call has an implicit delegated constructor call:
```kt
class A {
  constructor(i : Int) // implicit delegated constructor call
}
```
Resolving such case is added to FIR IDE HL API.

Still, rendered results mismatch, mainly because of resolution results. It was due to missed logic to resolve constructor call and return `PsiClass`. Thankfully, there is a util to convert `PsiType` to `PsiClass`, and FE 1.0 UAST also uses that for the same base service API. The counterpart in FIR UAST is added in this PR.

Finally, some constructors, in particular, in anonymous objects, are missing. It was fixed in FIR LC part.

### CODE REVIEW

**Comment by jsjeon** [comment]
Note that this one depends on [KOTLIN-CR-157](https://kotlin.jetbrains.space/p/kotlin/reviews/157/).

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt [hunk 1]
```diff
@@ -19,6 +25,14 @@ val firKotlinUastPlugin: FirKotlinUastLanguagePlugin by lz {
         ?: FirKotlinUastLanguagePlugin()
 }
 
+internal fun KtType.toPsiClass(context: KtElement): PsiClass? {
```
Comment (darthorimar) [inline uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: `KtType` should never be passed as receiver and leaked outside analyse call. Please, consider rewriting the code as the following:

```
internal fun KtAnalysisSession.toPsiClass(type: KtType, context: KtElement): PsiClass? {
      return asPsiType(context, TypeMappingMode.DEFAULT_UAST)?.let {
            PsiTypesUtil.getPsiClass(it)
    }
}

```
and always call `toPsiClass` insdie the `analyseForUast` lambda

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt [hunk 2]
```diff
@@ -33,6 +37,19 @@ internal fun KtType.toPsiClass(context: KtElement): PsiClass? {
     }
 }
 
+internal fun KtCallElement.toFunctionLikeSymbol(): KtFunctionLikeSymbol? {
```
Comment (darthorimar) [inline uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: The same as in https://github.com/JetBrains/intellij-kotlin/pull/96#discussion_r684035608 . `KtSymbol` should not be leaked outside `analyse` call

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt [hunk 3]
```diff
@@ -33,6 +37,19 @@ internal fun KtType.toPsiClass(context: KtElement): PsiClass? {
...
+    analyseForUast(this) {
+        return resolveCall()?.targetFunction?.candidates?.singleOrNull()
+    }
+}
+
+internal fun KtCallElement.toAnnotationConstructorSymbol(): KtConstructorSymbol? {
```
Comment (darthorimar) [inline uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: The same as in https://github.com/JetBrains/intellij-kotlin/pull/96#discussion_r684035608 . `KtSymbol` should not be leaked outside `analyse` call (2)

**Review by darthorimar — CHANGES_REQUESTED** [review]

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt (hunk 1 above)
Comment (jsjeon) [reply uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: Good to learn that. Rewritten as per the suggestion, thanks.

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt (hunk 2 above)
Comment (jsjeon) [reply uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: I can't come up with a better util signature, other than leaking `KtSymbol` :p So, it looks not quite necessary; I dropped the resolution utilizing commit as a whole.

File: uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt (hunk 3 above)
Comment (jsjeon) [reply uast/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/internal/firKotlinInternalUastUtils.kt]: Ditto: dropped.

**Comment by darthorimar** [comment]
LGTM 👍 merged into https://github.com/JetBrains/intellij-community/tree/kt-212-master

**Pull request closed without merging**